src/apple/macos/ffi.rs

```diff
@@ -154,3 +177,133 @@ mod io_service {
+    
+    pub fn IOHIDEventSystemClientSetMatching(client: *mut libc::c_void, matches: CFDictionaryRef) -> i32;
+
+    pub fn IOHIDEventSystemClientCopyServices(client: *mut libc::c_void) -> CFArrayRef;
+
+    pub fn IOHIDServiceClientCopyProperty(service: IOHIDServiceClientRef, key: CFStringRef) -> CFStringRef;
+
+    pub fn IOHIDServiceClientCopyEvent(service: IOHIDServiceClientRef, v0: i64, v1: i32, v2: i64) -> IOHIDEventRef;
+
+    pub fn IOHIDEventGetFloatValue(event: IOHIDEventRef, field: i64) -> f64;
+}
+
+pub type CFMutableArrayRef = *mut __CFArray;
+
+extern "C" {
+    pub fn CFArrayCreateMutable(allocator: CFAllocatorRef, capacity: CFIndex, callbacks: *const CFArrayCallBacks) -> CFMutableArrayRef;
+
+    pub fn CFArrayAppendValue(the_array: CFMutableArrayRef, value: *const libc::c_void);
+}
+
+pub(crate) const HID_DEVICE_PROPERTY_VENDOR_ID: &str = "VendorId";
+pub(crate) const HID_DEVICE_PROPERTY_PRODUCT_ID: &str = "ProductID";
+pub(crate) const HID_DEVICE_PROPERTY_PRODUCT: &str = "Product";
+pub(crate) const HID_DEVICE_PROPERTY_PRIMARY_USAGE: &str = "PrimaryUsage";
+pub(crate) const HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE: &str = "PrimaryUsagePage";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_INPUT_REPORT_SIZE: &str = "MaxInputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_MAX_OUTPUT_REPORT_SIZE: &str = "MaxOutputReportSize";
+pub(crate) const HID_DEVICE_PROPERTY_REPORT_ID: &str = "ReportID";
+
+pub(crate) const kHIDPage_AppleVendor: i32 = 0xff00;
+pub(crate) const kHIDUsage_AppleVendor_TemperatureSensor: i32 = 0x0005;
+
+pub(crate) fn matching(page: i32, usage: i32) -> CFDictionaryRef {
+    unsafe {
+        let primary_usage_page = CString::new(HID_DEVICE_PROPERTY_PRIMARY_USAGE_PAGE).unwrap();
+        let primary_usage = CString::new(HID_DEVICE_PROPERTY_PRIMARY_USAGE).unwrap();
+
+        let keys = [
+            CFStringCreateWithCString(
+                null() as *const _, 
+                primary_usage_page.as_ptr(), 
+                0
+            ),
+            CFStringCreateWithCString(
+                null() as *const _, 
+                primary_usage.as_ptr(), 
+                0
+            ),
+        ];
+
+        let nums = [
+            CFNumberCreate(
+                null(), 
+                kCFNumberSInt32Type, 
+                &page as *const _ as *const libc::c_void
+            ),
+            CFNumberCreate(
+                null(), 
+                kCFNumberSInt32Type, 
+                &usage as *const _ as *const libc::c_void
+            ),
+        ];
+
+        CFDictionaryCreate(
+            null(), 
+            &keys as *const _ as *const _,
+            &nums as *const _ as *const _, 
+            2,
+            &kCFTypeDictionaryKeyCallBacks,
+            &kCFTypeDictionaryValueCallBacks
+        )
+    }
+}
+
+pub fn get_product_names(sensors: CFDictionaryRef) -> Vec<String> {
+    unsafe {
+        let system = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
```

Comment:
Please check all pointers here. If any pointer is null, I assume the function should stop right away.